text/0000-fabric-private-chaincode.md

```diff
@@ -0,0 +1,472 @@
+
+* Step 3: Endorsement
+
+	![Endorsement](../images/fpc/high-level/FPC-Endorsement.png)
+
+	The Client receives the Proposal Response (and collects enough Proposal Responses from other endorsing peers to satisfy the chaincode’s Endorsement Policy). Moreover, the FPC Client SDK extension verifies that the proposal response signature has been produced by a "valid" Chaincode Enclave. If this verification step fails, there is no need for the client to proceed and the transaction invocation is aborted. Otherwise, the client continues and builds a transaction and submits it for ordering.
+
+* Step 4: Ordering
+
+	![Ordering](../images/fpc/high-level/FPC-Ordering.png)
+
+	With FPC we follow the normal Ordering service. While the Orderer is unmodified in FPC, the ordered transactions broadcast to the Peers in the Channel for validation now include attested endorsements. Note that we encourage the use of a BFT-based ordering service as FPC requires the Ordering service to be trusted.
+
+* Step 5: Validation
+
+	![Validate](../images/fpc/high-level/FPC-Validate.png)
+
+	As the Peers in the Channel receive the block of transactions, they perform the standard Fabric validation process.
+	In addition, a custom validation plugin in the Peer is responsible to verify FPC transactions. In particular, the FPC Validator queries its local FPC Registry to retrieve the enclave signature verification key and check that the signature was produced in an actual and correct Chaincode Enclave. This query retrieves the stored Attestation report associated with the Public Key of the Chaincode Enclave that produced and signed the transaction. The Attestation report is checked to verify the details of the Chaincode Enclave, affirming the validity of the Enclave.
+
+* Step 6: Trusted Ledger Revalidation
+
+	![Revalidate](../images/fpc/high-level/FPC-Revalidate.png)
+
+	In addition to the standard validation and commit process, the Peer also forwards all committed blocks to the Ledger Enclave in order to establish a full and current trusted view of the ledger. 
+	The same validation steps described above are repeated inside the Enclave, and then the transaction is committed to the trusted version of the Ledger inside this Enclave. This constitutes an update to the World State Integrity Metadata: for each Key Value Pair of World State, a second Key Value Pair is stored in the Trusted Ledger containing Integrity Metadata (a cryptographic hash value) along with channel-specific details needed to verify that the transaction was produced by a valid authorized participant in the channel.
```

Comment:
> for each Key Value Pair of World State, a second Key Value Pair is stored in the Trusted Ledger containing Integrity Metadata (a cryptographic hash value) along with channel-specific details needed to verify that the transaction was produced by a valid authorized participant in the channel.

So the enclave has limited storage capabilities that can securely store a hash, then? Why not instead use a MAC and just MAC the hash and channel pair? Then you can use this MAC key for all channels and you don't need to store in the enclave a value for each channel?

Reply:
currently, the meta-data must fit into enclave memory. Note though, even with currently still somewhat limited EPC, this supportes millons of keys.  As noted below, we also have it on the road-map to do crypto-paging, which besides scaling the number of possible keys also improves start-up performance for peers.

Reply:
> the Peer also forwards all committed blocks to the Ledger Enclave in order to establish a full and current trusted view of the ledger. 

I don't understand something. 
It seems to me that the Ledger Enclave needs to process all blocks, and validate the transactions in these blocks, namely configuration transactions. 

I believe this is needed, because otherwise the Enclave Ledger cannot maintain the correct view of the validity of the transactions inside of it.

Under the assumption that I'm correct and it is indeed needed, it seems to me that the Ledger Enclave needs to implement **all logic** of transaction validation: **everything** that Fabric transaction validation depends on, namely MSP, BCCSP, and config transaction processing. 
Otherwise, a malicious party can look at the ledger enclave code and at the Fabric code, see a deviation in the semantics and craft a transaction that is invalid in Fabric but valid in the Enclave code, and thus cause a fork between the Fabric state and the Ledger Enclave state. Am I wrong?

If that is indeed the case, isn't that a lot of code to implement and maintain in C++ ? Fabric code has historically mutated across versions, and a lot of Fabric versions come with activation flags (called "Capabilities") which hint the validation pipeline whether they should be applied or not. Maintaining this code in Fabric has been an unpleasant and time consuming endeavor and you can look that we have validation code duplicated for [three](https://github.com/hyperledger/fabric/tree/a282f7132565d509d3ba6d502cd9c07cc5e567f0/core/handlers/validation/builtin/v12) [different](https://github.com/hyperledger/fabric/tree/a282f7132565d509d3ba6d502cd9c07cc5e567f0/core/handlers/validation/builtin/v13) [versions](https://github.com/hyperledger/fabric/tree/a282f7132565d509d3ba6d502cd9c07cc5e567f0/core/handlers/validation/builtin/v20).

Reply:
@yacov You are right, re-implementing the entire validation pipeline of the peer in C++ and run it inside the ledger enclave makes maintaining the code not easy.

I opened an [issue](https://github.com/hyperledger-labs/fabric-private-chaincode/issues/402) for that  topic. So we can have find a solution/strategy in a separate conversation without polluting this conversation here.

Reply:
Right, I think the answer really lies in reducing the scope and perhaps even deviating from the current Fabric logic in the FPC, to eventually reach something that co-exists within Fabric and appears to the user as Fabric, but has a more strict (but simple) validation logic than Fabric.